File: src/4장/4-3.py
def Brute_Force(idx, x, y):
    global answer
    if(idx == m):
        choice_chicken = []
        for i in range(n):
            for j in range(n):
                if(City[i][j] == 3):
                    choice_chicken.append((i,j))
        res = Min_Distance(choice_chicken, house)
        if(answer > sum(res)):
            answer = sum(res)
        return
    else:
        for i in range(x, n):
            if(i == x): k = y
            else: k = 0
            for j in range(k, n):
                if(City[i][j] == 2):
                    City[i][j] = 3
                    Brute_Force(idx+1, i, j+1)
                    City[i][j] = 2

def Min_Distance(chicken,house):
    sum_Distance = []
    for i in house:
        min_D = 987654321
        for j in chicken:
            Distance = abs(i[0]-j[0]) + abs(i[1]-j[1]) 
            min_D = min(min_D, Distance)
        sum_Distance.append(min_D)
    return sum_Distance

n, m = map(int, input().split())
City = [list(map(int, input().split())) for _ in range(n)]
answer = 987654321
house = []
for i in range(n):
    for j in range(n):
        if(City[i][j] == 1):
            house.append((i,j))
Brute_Force(0,0,0)
print(answer)

File: src/4장/4-3.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
	static int N, M, answer = Integer.MAX_VALUE;
	static boolean[] choice_chicken;
	static List<int[]> house, chicken;

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		M = sc.nextInt();
		house = new ArrayList<>();
		chicken = new ArrayList<>();

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				int tmp = sc.nextInt();
				if (tmp == 1)
					house.add(new int[] { i, j });
				else if (tmp == 2)
					chicken.add(new int[] { i, j });
			}
		}
		choice_chicken = new boolean[chicken.size()];
		Brute_Force(-1, 0);

		System.out.println(answer);
	}

	static void Brute_Force(int idx, int cnt) {
		if (cnt == M) {
			int dist = 0;

			for (int[] h : house) {
				int tmp = Integer.MAX_VALUE;
				for (int i = 0; i < choice_chicken.length; i++) {
					if (choice_chicken[i])
						tmp = Math.min(tmp, Math.abs(h[0] - chicken.get(i)[0]) + Math.abs(h[1] - chicken.get(i)[1]));
				}
				dist += tmp;
			}
			answer = Math.min(answer, dist);
			return;
		}

		for (int i = idx + 1; i < choice_chicken.length; i++) {
			choice_chicken[i] = true;
			Brute_Force(i, cnt + 1);
			choice_chicken[i] = false;
		}
	}
}